vhdx/src/vhdx_metadata.rs

```diff
@@ -0,0 +1,285 @@
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct DiskSpec {
+    disk_id: u128,
+    block_size: u32,
+    sectors_per_block: u32,
+    virtual_disk_size: u64,
+    logical_sector_size: u32,
+    physical_sector_size: u32,
+    chunk_ratio: u64,
+}
+
+struct MetadataTableHeader {
+    signature: u64,
+    reserved: u16,
+    entry_count: u16,
+    reserved2: [u8; 20],
+}
+
+impl MetadataTableHeader {
+    pub fn new(buffer: &[u8]) -> Result<MetadataTableHeader> {
+        let signature = LittleEndian::read_u64(&buffer[0..8]);
+        if signature != METADATA_SIGN {
+            return Err(Error::InvalidMetadataSign);
+        }
+
+        let reserved = LittleEndian::read_u16(&buffer[8..10]);
+        let entry_count = LittleEndian::read_u16(&buffer[10..12]);
+        if entry_count > METADATA_MAX_ENTRIES {
+            return Err(Error::InvalidEntryCount);
+        }
+
+        Ok(MetadataTableHeader {
+            signature,
+            reserved,
+            entry_count,
+            reserved2: buffer[12..32]
+                .try_into()
+                .expect("slice with incorrect length"),
+        })
+    }
+}
+
+pub struct MetadataTableEntry {
+    item_id: u128,
+    offset: u32,
+    length: u32,
+    flag_bits: u32,
+    reserved: u32,
+}
+
+impl MetadataTableEntry {
+    fn new(buffer: &[u8]) -> Result<MetadataTableEntry> {
+        let item_id = LittleEndian::read_u128(&buffer[0..16]);
+        let offset = LittleEndian::read_u32(&buffer[16..20]);
+        let length = LittleEndian::read_u32(&buffer[20..24]);
+        let flag_bits = LittleEndian::read_u32(&buffer[24..28]);
+        let reserved = LittleEndian::read_u32(&buffer[28..32]);
+
+        Ok(MetadataTableEntry {
+            item_id,
+            offset,
+            length,
+            flag_bits,
+            reserved,
+        })
+    }
+
+    pub fn parse_metadata(f: &mut File, metadata_region: RegionTableEntry) -> Result<DiskSpec> {
+        let mut disk_spec: DiskSpec = { Default::default() };
+        let mut metadata_presence: u16 = 0;
+
+        let mut buffer = [0u8; METADATA_TABLE_MAX_SIZE];
+        f.seek(SeekFrom::Start(metadata_region.file_offset))
+            .map_err(Error::ReadMetadata)?;
+        f.read_exact(&mut buffer).map_err(Error::ReadMetadata)?;
+
+        let mut header_slice = [0u8; METADATA_ENTRY_SIZE];
+        header_slice.copy_from_slice(&buffer);
+
+        let metadata_header = match MetadataTableHeader::new(&header_slice) {
+            Ok(metadata_header) => metadata_header,
+            Err(e) => {
+                return Err(e);
+            }
+        };
+
+        let offset = METADATA_ENTRY_SIZE;
+        for _ in 0..metadata_header.entry_count {
+            let mut entry_slice = [0u8; METADATA_ENTRY_SIZE];
+
+            entry_slice.copy_from_slice(&buffer[offset..offset + METADATA_ENTRY_SIZE]);
+            let metadata_entry = match MetadataTableEntry::new(&entry_slice) {
```

Comment:
Same here, I don't think you need/should copy.

Reply:
Done.